crates/ra_assists/src/handlers/extract_struct_from_enum_variant.rs

```diff
@@ -0,0 +1,331 @@
+use ra_ide_db::{
+    defs::Definition, imports_locator::ImportsLocator, search::Reference, RootDatabase,
+};
+use ra_syntax::{
+    algo::find_node_at_offset,
+    ast::{self, AstNode, NameOwner},
+    SourceFile, SyntaxNode, TextRange, TextSize,
+};
+
+use crate::{
+    assist_context::{AssistBuilder, AssistDirector},
+    utils::insert_use::insert_use_statement_with_string_path,
+    AssistContext, AssistId, Assists,
+};
+use ast::{ArgListOwner, VisibilityOwner};
+use hir::{EnumVariant, Module, ModuleDef};
+use ra_db::FileId;
+use ra_fmt::leading_indent;
+use rustc_hash::FxHashSet;
+
+// Assist: extract_struct_from_enum_variant
+//
+// Extracts a struct from enum variant.
+//
+// ```
+// enum A { <|>One(u32, u32) }
+// ```
+// ->
+// ```
+// struct One(pub u32, pub u32);
+//
+// enum A { One(One) }
+// ```
+pub(crate) fn extract_struct_from_enum_variant(
+    acc: &mut Assists,
+    ctx: &AssistContext,
+) -> Option<()> {
+    let variant = ctx.find_node_at_offset::<ast::EnumVariant>()?;
+    let field_list = match variant.kind() {
+        ast::StructKind::Tuple(field_list) => field_list,
+        _ => return None,
+    };
+    let variant_name = variant.name()?.to_string();
+    let variant_hir = ctx.sema.to_def(&variant)?;
+    if existing_struct_def(ctx.db, &variant_name, &variant_hir) {
+        return None;
+    }
+    let enum_ast = variant.parent_enum();
+    let enum_name = enum_ast.name()?.to_string();
+    let visibility = enum_ast.visibility();
+    let current_module_def =
+        ImportsLocator::new(ctx.db).find_imports(&enum_name).first()?.left()?;
+    let current_module = current_module_def.module(ctx.db)?;
```

Comment:
Ah, that's a pretty round-about way to get from syntax to def :-) There's `sema.to_def()` for this purpose:

```rust
    let current_module_def: hir::Enum = ctx.sema.to_def(&enum_ast)?;
    let current_module: hir::Module = current_module_def.module(ctx.db);
```

Reply:
Yes, thanks!